File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.cs
using System;
using System.Collections.Generic;

public class Solution {
    public string[] solution(string[] players, string[] callings) {
        var pos = new Dictionary<string, int>(players.Length);
        for(int i = 0; i < players.Length; i++)
        {
            pos[players[i]] = i;
        }
        
        foreach(var name in callings)
        {
            int i = pos[name];
            
            string ahead = players[i - 1];
            players[i - 1] = name;
            players[i] = ahead;
            
            pos[name] = i - 1;
            pos[ahead] = i;
        }
        return players;
    }
}

File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.cpp
#include <string>
#include <vector>
#include <bits/stdc++.h>

using namespace std;

vector<string> solution(vector<string> players, vector<string> callings) {
    unordered_map<string, int> rank;
    
    for(int i = 0; i < players.size(); i++)
    {
        rank[players[i]] = i;
    }
    
    for(string call : callings)
    {
        int idx = rank[call];
        string front = players[idx - 1];
        
        swap(players[idx], players[idx - 1]);
        
        rank[call]--;
        rank[front]++;
    }
    return players;
}